### PR TITLE
property: use a stack to efficiently convert index to string

### LINKED ISSUE #15867: ossl_property_list_to_string() is a very heavy operation
This function call is computationally quite expensive.  It should be streamlined.

The limiting factor is search for the string by index -- this should be done using an array not by iterating through the entire table of strings.  The indices are dense and start from 1.

This wasn't a problem when the conversion back to string was a purely diagnostic operation but now it is used in the child library context code too.

### PR DESCRIPTION
**Pull request by paulidale**
The existing code does this conversion by searching the hash table for the appropriate index which is slow and expensive.  Also add some low level unit tests for the `ossl_property_name_str` and `ossl_property_value_str` functions.
    
Fixes #15867

- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: crypto/property/property_string.c [hunk 1]
```diff
@@ -134,91 +136,87 @@ static PROPERTY_STRING *new_property_string(const char *s,
...
+                property_free(ps_new);
+                CRYPTO_THREAD_unlock(propdata->lock);
+                return 0;
+            }
             lh_PROPERTY_STRING_insert(t, ps_new);
             if (lh_PROPERTY_STRING_error(t)) {
+                /*-
+                 * Deleting this entry will break the lock step between the
+                 * index and the stack.  Instead we zap the pointer and
+                 * refer to something that doesn't need to be freed.
```
Comment (t8m) [inline crypto/property/property_string.c:180]: Couldn't we just decrement *pidx? We are under lock.

Comment (paulidale) [reply crypto/property/property_string.c:180]: Yes, this is a better way to achieve the outcome.

**Review by t8m — APPROVED** [review]

**Comment by paulidale** [comment]
Merged to both.

I intended to add some small memory conditionals but will now do so in a different PR.

**Comment by t8m** [comment]
> Merged to both.
> 
> I intended to add some small memory conditionals but will now do so in a different PR.

@paulidale but this was approved for master branch only as this is a feature. IMO this should be reverted on 3.0 branch or an explicit exception has to be given by OTC.

**Comment by paulidale** [comment]
Bugger, you are correct.  This needs to be reverted.

**Comment by paulidale** [comment]
Revert in #17403

**Pull request closed without merging**